### DISCUSSION TITLE
How to get base ref branch in a GitHub Actions workflow?

### DISCUSSION THREAD

**Discussion by merc1er** [community/community#26243] [Actions]
<p>I want to know against which base a branch has a pull request open (if any). For example I have a <code>dev</code> branch and I want to perform a certain job <strong>ONLY</strong> if it has a PR open against <code>master</code>.</p>
<p>In the <a href="https://docs.github.com/en/actions/learn-github-actions/contexts" rel="noopener nofollow ugc">docs</a>, it says we can see the base ref with <code>github.base_ref</code>. But it also says:</p>
<blockquote>
<p>This property is only available when the event that triggers a workflow run is either pull_request or pull_request_target.</p>
</blockquote>
<p>Is there a way to know the base ref on every push to <code>dev</code>?</p>
<p>This is my code, but this only works once when I create the PR (and not on every push):</p>
<pre><code class="lang-yaml">jobs:
  my-job:
    runs-on: ubuntu-latest
    if: ${{ github.base_ref == 'refs/heads/dev' }}
    ...
</code></pre>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="merc1er" data-post="1" data-topic="262563">
<div class="title">
<div class="quote-controls"></div>
 merc1er:</div>
<blockquote>
<p>Is there a way to know the base ref on every push to <code>dev</code>?</p>
</blockquote>
</aside>
<p>No, because a push doesn’t have a “base ref”, just a before and after state on the branch that’s being pushed.</p>
<p>If you want to act based on a PR you need to use one of the PR-related events (or iterate over PRs using the API, but that could get messy and run afoul of rate limits).</p>

**Reply by thalysonalexr** [reply]
why not have?
in my local machine i do checkout in branch X, i pushed info for github. Github not storage this information anywhere local?

**Reply by airtower-luna** [reply]
It's because there is no "base". In a PR you have a "head branch" which you want to merge into a "base branch". Their respective refs are called "head ref" and "base ref".

In a push there is only one branch. So the push event simply has the information that the branch was at commit A before the push, and at commit B after.

**Reply by jsoref** [reply]
Git is structured around commits.
Tags and branches are just pointers to commits.

Most other things are educated guesses.

If someone tells git "change this branch from pointing at commit A to pointing to commit B", then people can reasonable conclude that the branch moved from A to B.

The basic way this happens is:
```sh
git checkout branch-name
touch file
git add file
git commit -m 'Adding file'
```

Here, we've added a commit on top of another commit. Obviously it's connected to the previous commit.

It could also happen like this:
```sh
git checkout branch-name
git reset --hard $COMMIT_B
```

In this case, `$COMMIT_B` isn't necessarily connected to `$COMMIT_A`.


But, you could have done this locally:

```sh
git checkout HEAD~
git branch -D branch-name
git branch -f branch-name $COMMIT_B
```

In both cases there's no actual relation between `$COMMIT_A` and `$COMMIT_B`, and in the second case, there isn't even a continuity between `branch-name` at `$COMMIT_A` and `branch-name` at `$COMMIT_B`.
But, because when you do a push to github, it _knows_ that `branch-name` used to be at `$COMMIT_A`, it can say "well, `branch-name` used to be `$COMMIT_A`"

Note that it has no idea what branch `$COMMIT_B` came from. There could be 0, 2, 3,  or more branches pointing to `$COMMIT_B`. It's also possible there are 0, 4, or 7 tags pointing at `$COMMIT_B`.

You could also do:

```sh
git branch -f branch-name some-tag~5
```

In which case `branch-name` is now a distant relative of `some-tag` but not actually directly related to it.

Or, you could do:

```sh
git checkout branch-a
git merge branch-b -m 'merging branch-a and branch-b'
git branch -f branch-name
```

Here, `branch-name` is now related to both `branch-a` and `branch-b`, but not equal to either.

In each case, you can `git push -f origin branch-name` and github will have to offer answers.

**Reply by thalysonalexr** [reply]
thanks so much for perfect and detailed answer

**Comment by jsoref** [comment]
<p>So, I have special code that looks to see if a git push has a corresponding open PR:</p><aside class="onebox githubblob" data-onebox-src="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L41-L47">
  <header class="source">

      <a href="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L41-L47" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L41-L47" target="_blank" rel="noopener">check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L41-L47</a></h4>



    <pre class="onebox"><code class="lang-sh">
      <ol class="start lines" start="41" style="counter-reset: li-counter 40 ;">
          <li>if to_boolean "$INPUT_SUPPRESS_PUSH_FOR_OPEN_PULL_REQUEST"; then</li>
          <li>  pull_request_json=$(mktemp_json)</li>
          <li>  pull_request_headers=$(mktemp)</li>
          <li>  pull_heads_query="$GITHUB_API_URL/repos/$GITHUB_REPOSITORY/pulls?head=${GITHUB_REPOSITORY%/*}:$GITHUB_REF"</li>
          <li>  keep_headers=1 call_curl \</li>
          <li>    "$pull_heads_query" &gt; $pull_request_json</li>
          <li>  mv "$response_headers" "$pull_request_headers"</li>
      </ol>
    </code></pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<aside class="onebox githubblob" data-onebox-src="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L60-L66">
  <header class="source">

      <a href="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L60-L66" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L60-L66" target="_blank" rel="noopener">check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L60-L66</a></h4>



    <pre class="onebox"><code class="lang-sh">
      <ol class="start lines" start="60" style="counter-reset: li-counter 59 ;">
          <li>elif [ $(jq length $pull_request_json) -gt 0 ]; then</li>
          <li>  (</li>
          <li>    open_pr_number=$(jq -r '.[0].number' $pull_request_json)</li>
          <li>    echo "Found [open PR #$open_pr_number]($GITHUB_SERVER_URL/$GITHUB_REPOSITORY/$open_pr_number) - check-spelling should run there."</li>
          <li>    echo</li>
          <li>    echo '::warning ::WARNING: This workflow is intentionally terminating early with a success code -- it has not checked for misspellings.'</li>
          <li>    echo '::warning ::You should treat this workflow run as if it were SKIPPED state and instead look for a `pull_request_target` workflow for `check-spelling` in the PR.'</li>
      </ol>
    </code></pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<aside class="onebox githubblob" data-onebox-src="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L74-L75">
  <header class="source">

      <a href="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L74-L75" target="_blank" rel="noopener">github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://github.com/check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L74-L75" target="_blank" rel="noopener">check-spelling/check-spelling/blob/624e4ede7c93c6255ad80c08395fc1687b6a3e10/unknown-words.sh#L74-L75</a></h4>



    <pre class="onebox"><code class="lang-sh">
      <ol class="start lines" start="74" style="counter-reset: li-counter 73 ;">
          <li>  exit 0</li>
          <li>fi</li>
      </ol>
    </code></pre>



  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>You can do all sorts of magical things. Whether you should is a different story. In my case, I like that my action/workflow can run on both <code>push</code> and <code>pull_request</code>(<code>_target</code>) and not cost too much extra/cause too much extra confusion if both are configured.</p>

**Comment by tony9823** [comment]
To check if a branch (e.g., dev) has an open pull request against a base branch (e.g., master) on every push, you can use the GitHub API in your workflow. Here's how:

Use the on: push event for the dev branch in your workflow [file](https://grilldmenu.com/).
Use the GitHub REST API to query open pull requests where the head is dev and the base is master.
Here's an example GitHub Actions workflow:

yaml
Copy
Edit
name: Check PR Base Ref

on:
  push:
    branches:
      - dev

jobs:
  check-pr:
    runs-on: ubuntu-latest
    steps:
      - name: Check if PR is open against master
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
        run: |
          PR_RESPONSE=$(curl -s -H "Authorization: token $GITHUB_TOKEN" \
            "https://api.github.com/repos/${{ github.repository }}/pulls?head=${{ github.repository_owner }}:dev&base=master&state=open")
          if [ "$PR_RESPONSE" = "[]" ]; then
            echo "No open PR against master."
            exit 0
          else
            echo "PR open against master. Proceeding with the job."
          fi
Explanation:
Trigger on Push: The workflow triggers on any push to the dev branch.
GitHub API Request: Queries all pull requests with the head as dev and the base as master.
Decision: If no PR is found ([]), the job stops; otherwise, it proceeds.
